### ISSUE TITLE
[FR] Is it possible to have an Rcpp interface?

### ISSUE THREAD

**Issue by shrektan** [astamm/nloptr#39]
Hi, @jyypma , as we all know, there're overheads to call R functions. `nloptr` uses pure R functions as it's target function may jeopardize the performance for large-scale computing. 

`nlopt` itself contains a c++ interface, in the header `nlopt.hpp`.  So I'm asking if there's a way to share the header and provide a linkage so that we can write some nice Rcpp codes like the following example. 

Thanks.

(Here's a simple use example I take from https://nlopt.readthedocs.io/en/latest/NLopt_Tutorial/#example-in-c with slight modification)

```cpp
#include <Rcpp.h>
#include <nlopt.hpp>

// [[Rcpp::plugins(cpp11)]]
// [[Rcpp::depends(nloptr)]]
struct my_constraint_data
{
    double a;
    double b;
};

double myvfunc(const std::vector<double> &x, std::vector<double> &grad, void *my_func_data)
{
    if (!grad.empty()) {
        grad[0] = 0.0;
        grad[1] = 0.5 / sqrt(x[1]);
    }
    return sqrt(x[1]);
}


double myvconstraint(const std::vector<double> &x, std::vector<double> &grad, void *data)
{
    my_constraint_data *d = reinterpret_cast<my_constraint_data*>(data);
    double a = d->a, b = d->b;
    if (!grad.empty()) {
        grad[0] = 3 * a * (a*x[0] + b) * (a*x[0] + b);
        grad[1] = -1.0;
    }
    return ((a*x[0] + b) * (a*x[0] + b) * (a*x[0] + b) - x[1]);
}

// [[Rcpp::export]]
double test()
{
    nlopt::opt opt(nlopt::LD_MMA, 2);
    std::vector<double> lb(2);
    lb[0] = R_NegInf;
    lb[1] = 0.0;
    opt.set_lower_bounds(lb);
    opt.set_min_objective(myvfunc, NULL);
    my_constraint_data data[2] = { {2,0}, {-1,1} };
    opt.add_inequality_constraint(myvconstraint, &data[0], 1e-8);
    opt.add_inequality_constraint(myvconstraint, &data[1], 1e-8);
    
    opt.set_xtol_rel(1e-4);
    std::vector<double> x(2);
    x[0] = 1.234; 
    x[1] = 5.678;
    double minf;
    nlopt::result result = opt.optimize(x, minf);
    Rcpp::Rcout << int(result) << "\n";
    return minf;
}
```

**Comment by jyypma** [comment]
Yes, that should be possible and is something I've started working on
during recent commits.

**Comment by shrektan** [comment]
Well thanks. I’m looking forward to that.

**Comment by eddelbuettel** [comment]
I'll try to help. 

I got contacted by a user of RcppArmadillo and nloptr who has a hard time making them work together.

**Comment by jyypma** [comment]
Since a couple of days the master version on github now exposes the NLopt C
functions through the nloptrAPI.h header file. An example of how this works
based on the C code in the NLopt tutorial can be found in one of the unit
tests:
https://github.com/jyypma/nloptr/blob/master/tests/testthat/test-C-api.R
Since yesterday both builds pass on Travis (both the one linking to an
external NLopt library and the one where NLopt is compiled during the
configuration of nloptr). The build created by WinBuilder also seemed to
work fine.

There is no support for Rcpp as I'm not yet sure what would be needed for
that, but any advice from Dirk on that front is of course welcomed!



On Tue, Oct 10, 2017 at 4:41 PM, Dirk Eddelbuettel <notifications@github.com

**Comment by eddelbuettel** [comment]
That is awesome and maybe all my collaborator needs.  Does this code then resolve against the library your R package provides, ie an R user would not need anything else?

(Getting something working first is best, we can always make it "pretty" or "user-friendly" afterwards.)

**Comment by jyypma** [comment]
Yes, as far as I could test, an R user would not have to manually install
NLopt (but still can if that's preferred). The necessary flags can be
obtained with nloptr:::CFlags() and nloptr::LdFlags(). As far as I
understand nloptr.[so|dll] will automcatically be available after
library(nloptr), so I'd think that the linker flags are only necessary in
case nloptr links against an external NLopt library. I'm not sure though
whether a package should import or depend on nloptr and if the NAMESPACE
file needs some adjustment as well.

**Comment by daijiang** [comment]
Is there any code examples about how to use this with RcppArmadillo? The code in the test folder is helpful. But I don't know how to put the code provided at the beginning of this issue as a script, then use rcpp to source it so that I can use the function in R? 

When I copy the code in the test file into a c++ file in RStudio, it said `nlopt_opt` is an unknown type.

Thanks. @eddelbuettel @jyypma @shrektan

**Comment by shrektan** [comment]
@daijiang I create a simple repo https://github.com/shrektan/rcpp_nlopt_example based on my original example. However, I don't have time to include `RcppArmadillo` and you need to figure out by yourself. 

Moreover, if the system is not windows x64, you have to change the linking library and `dyn.load()` may no longer needed.

**Comment by daijiang** [comment]
@jyypma Thanks for the repo. I will check it out.

**Comment by daijiang** [comment]
I still did not figure out how to use the c++ files directly in an R package even with help from @shrektan . I wonder can any of you have a look at this file ([https://github.com/daijiang/phyr/blob/master/src/pglmm_gaussian.cpp#L267-L283](https://github.com/daijiang/phyr/blob/master/src/pglmm_gaussian.cpp#L267-L283))? The current implement imports the `nloptr` function into c++, which probably is time costly. 

What I want to do is to link to the header files from `nloptr` and then can start use the `nlopt` function in the c++ file. But at this moment, I have no clue where to start since I have very limited experience with c++. Thank you for your help. @jyypma @eddelbuettel

**Comment by jchiquet** [comment]
Hi and many thanks @jyypma for the great work on nloptr.

I would also enjoy using the Rcpp/RcppArmadillo features with the nlopt library without suffering the overheads to call R functions. I implemented two very basic R packages that use the C and C++ API of nlopt via nloptr and also link to Rcpp/RcppArmadillo : 

https://github.com/jchiquet/RcppArmadilloNLoptExample

In the first package nloptC I use the C API (building on the example of @jyypma (https://github.com/jyypma/nloptr/blob/master/tests/testthat/test-C-api.R) and everything works fine. In this case, I link to Rcpp, RcppArmadillo and nloptr in the DESCRIPTION file and the head of the cpp file in the src of my R package nloptC looks like

`#include "RcppArmadillo.h"`
`// [[Rcpp::depends(RcppArmadillo)]]`
`#include <nloptrAPI.h>`

However, when I try to use the C++ API like in the example of @daijiang (https://github.com/shrektan/rcpp_nlopt_example/blob/master/main.cpp) the package fails to load with error  `undefined` symbol: `nlopt_set_munge`.

The head of the cpp file ine the src of my R package nloptCpp looks like

`#include "RcppArmadillo.h"`
`// [[Rcpp::depends(RcppArmadillo)]]`
`#include <nloptrAPI.h>`
`#include <nlopt.hpp>`

Am I doing something wrong? Any help would be greatly appreciated. Thanks.

**Comment by eddelbuettel** [comment]
@jchiquet: Just _wow_ this is absolutely fabulous! 

Here is the *tl;dr* version: your example package in your example repo build just fine with minimal changes.

Longer version:
1. Again, thanks so much for doing this (and equal thanks of course to @jyypma for the added API)
2. Your are missing the link instruction to `libnlopt` in both your C example and your C++ example.
3. This bites you in C++ "because C++" but C lets you get way with it (on my Linux box, at least)
4. The fixes are small and we should make them.  I can send a PR.
5. We should also extend your example packages a little to run an example (as a test).
6. Your `DESCRIPTION` files also lack the `Imports:` implied by your `NAMESPACE`
7. For the C++ version I had to remove an export from the `NAMESPACE`.
8. For the C version we can export `std::vector<double>` directly, no need to `wrap()`

Again, this is all small fry.  The main key is that we are getting to nice, self-contained examples building on both nloptr and RcppArmadillo.  Many people wanted this for some time.

We will have a little more work to do (presumably on the side of nloptr) to create some helpers for use in `src/Makevars` to have nlopt exported.   And we may have _much more work to do_ to support the embedded nlopt library.  My use case here is external linking which is easier.  But we can likely export the entry points of nltop via nloptr to users such as RcppArmadilloNLoptExample.

Ok, pushed my changes to [my fork here](https://github.com/eddelbuettel/RcppArmadilloNLoptExample).

**Comment by jchiquet** [comment]
@eddelbuettel Happy to help ! It seems to me very, very little in return for my intensive use of nloptr, Rcpp, RcppArmadillo and others.

If you want to create a pull request from your fork, I can then add you as a collaborator of RcppArmadilloNLoptExample, your call.

Anyway, thanks for fixing this.

**Comment by eddelbuettel** [comment]
I already sent you a PR:  https://github.com/jchiquet/RcppArmadilloNLoptExample/pull/1

**Comment by daijiang** [comment]
From the users' point of view, is it possible to have a sugar function from nloptr (nloptcpp?) that will act in c++ as nloptr::nloptr acts in R? At this moment, users need to convert all control options from R to c++. It would be great to have a sugar function from nloptr to allow users do [these lines of code](https://github.com/jchiquet/RcppArmadilloNLoptExample/blob/master/nloptCpp/src/nlopt_cpp.cpp#L42-L57) with a simple function call in c++ with a list of control options? Something like:

`Rcpp::List results = nloptr::nloptcpp(x0, eval_f, options, ...)`

Or put it simper, can we replace the `nloptr` function (imported R function into c++) with `nloptcpp` (c++ sugar function?) in [these lines of code](https://github.com/daijiang/phyr/blob/master/src/pglmm_gaussian.cpp#L267-L283)?

This can be a huge help for users who want to use nlopt directly in c++. But it may need some work from the nloptr end. @jyypma @eddelbuettel @jchiquet

**Comment by eddelbuettel** [comment]
@jchiquet I by now have looked more closely at what Jelmer exports via `nloptrAPI.h`.  He is actually doing the right thing which means that we _can_ alter the example packages to explicitly import and not require external linking.  If you want an example of how that works you could look at my RcppXts package and xts, or a few other packages.   In short, the _client package_ needs to register the import when it loads.

*Edit:* Actually, we may be able to use it directly.  I guess I need to poke some...

**Comment by jchiquet** [comment]
@eddelbuettel ok sounds great. I will have a look at RcppXts asap and will try to follow the same approach.

**Comment by eddelbuettel** [comment]
It is actually much, much simpler, amd we really have two distinct examples.

- `nloptC` only use the properly exported API which Jelmer supports directly from `nloptr`; it does _not_ use any other symbols.  Ergo, no linking needed.  This should just work _everywhere_ which is really neat.  And nice to have an example.

- `nloptCpp` chooses to also use other symbols from the actual nlopt header.  Which unsurpisingly leads to the need to linking against the library to get these symbols. Fundamentally this _is_ a different approach and while more general it also more imposing.

**Comment by eddelbuettel** [comment]
@jchiquet I sent a second PR to your repo removing the unnecessary linking I added to your C example.

**Comment by jchiquet** [comment]
@eddelbuettel ok PR accepted thanks

**Comment by jchiquet** [comment]
Ok, I found a simple solution [for nlopCpp](https://github.com/jchiquet/RcppArmadilloNLoptExample/tree/master/nloptCpp) that uses the C++ interface of NLopt (following @shrektan's https://github.com/shrektan/rcpp_nlopt_example). Still, it bypasses nloptr (so I hope this post is still relevant here, probably at least for people looking for a solution).

I simply added the nlopt.hpp file in inst/include and use the following flags in the Makevars

`NLOPT_LIBS = -lnlopt -lm`
`NLOPT_CXXFlags = -I../inst/include`

It works on Linux Ubuntu 16.04 with libnlopt-dev installed.

At the moment, the C++ example in src/nlopt_cpp.cpp only uses some basics Rcpp features: I plain to write a inst/include/arma_nlopt.hpp that will essentially use `arma::vec` in place of `std::vector<double>` to call the nlopt C functions that perform the optimization.

**Comment by eddelbuettel** [comment]
Yes, that is exactly what we did already: 
https://github.com/eddelbuettel/RcppArmadilloNLoptExample/blob/master/nloptCpp/src/Makevars#L2-L3

As I tried to explain to you over those two PRs, the _C_ example works by only using what Jelmer exports.  This is easily portable and usable across CRAN as nloptr does the hard work of figuring out if system nlopt is available, or if a local embedded compilation has to be done.   The _C++_ example only works in the former case.   Nothing has changed: it needs system nlopt libraries.

**Comment by eddelbuettel** [comment]
That said we do need more examples, could do with a Rcpp Gallery post etc pp.  And of course for the C solution to be used more widely we need nloptr 1.1.0 released ...

**Comment by jchiquet** [comment]
@eddelbuettel thank you for the precision, and sorry for being a bit slow...

That being said, and for completeness, I needed -- at least on my system -  to include the nlopt.hpp file in inst/include and the additional flag

`NLOPT_CXXFlags = -I../inst/include`

on top of

`NLOPT_LIBS = -lnlopt -lm`

in order to avoid some sharing object error when loading the package nloptCpp.

**Comment by eddelbuettel** [comment]
Yes/no/maybe.  That is sloppy because you mix local per-package componentns (`nlopt.hpp`) with system-wide libraries (`-lnlopt` with a `-L....` flag).  Fine locally, not a good plan in a tutorial or widely used package.

The header should just be found.  When I sent you my PRs, nlopt headers were found just fine from the system wide location, ie 

```sh
edd@rob:~$ ls -l /usr/include/nlopt*
-rw-r--r-- 1 root root  3864 Jul  8  2017 /usr/include/nlopt.f
-rw-r--r-- 1 root root 13959 Jul  8  2017 /usr/include/nlopt.h
-rw-r--r-- 1 root root 21084 Jul  8  2017 /usr/include/nlopt.hpp
edd@rob:~$ 
```

There is never a need for setting `/usr/include`.

More generally, one relies on `pkg-config`:

```
edd@rob:~$ pkg-config --libs nlopt
-lnlopt -lm
edd@rob:~$ pkg-config --cflags nlopt

edd@rob:~$ 
```

And that is what our nloptr setup uses too.  We could probably export that with some help or interest from Jelmer.

**Comment by jchiquet** [comment]
Hum, I realized I had two conflicting installation of nlopt (one from nloptr and another one from libnlopt-dev) and was not pointing to the correct (or expected) header. So after purging + a clean installation of nlopt via libnlopt-dev, I do not need to include nlopt.hpp anymore in the package, nor `NLOPT_CXXFlags = -I../inst/include` of course. `NLOPT_LIBS = -lnlopt -lm` suffices.

Everything works fine just as Dirk said a week ago (at least the solution with a system nlopt installed), so thank again @eddelbuettel for your patience and for the pedagogical messages.

I updated (reverted) [RcppArmadilloNLoptExample/nloptCpp](https://github.com/jchiquet/RcppArmadilloNLoptExample/tree/master/nloptCpp) accordingly.

@jyypma Still looking for a solution that would let the possibility to use a local compilation of nlopt via nloptr, which would be easier for most users.

**Comment by jyypma** [comment]
What is the return value of nloptr:::CFlags() and nloptr:::LdFlags()? I
think you should be able to use those to use link against the NLopt library
embedded in nloptr. The LdFlags is empty I think when you have no external
NLopt. If you use library(nloptr) I could dynamically link against it, but
perhaps that is not how R intended it to work?

The NLopt test C API example works on Travis when there is no external
NLopt library installed. I tested the same on Windows and that also seems
to work for me.

**Comment by jchiquet** [comment]
Hi @jyypma thanks for helping.

If nloptr is installed on a system nlopt library, I get

``` R
> nloptr:::LdFlags()
 -lnlopt -lm> 
> nloptr:::CFlags()
-I"/home/jchiquet/R/x86_64-pc-linux-gnu-library/3.4/nloptr/include"> 
```

If no nlopt installation was found when installing nloptr , I get

``` R
> nloptr:::LdFlags()
 -lm> 
> nloptr:::CFlags()
-I"/home/jchiquet/R/x86_64-pc-linux-gnu-library/3.4/nloptr/include"> 
```
So I tried to use nloptr:::LdFlags() and nloptr:::CFlags() in the Makevars of  [RcppArmadilloNLoptExample/nloptCpp](https://github.com/jchiquet/RcppArmadilloNLoptExample/tree/master/nloptCpp), 

``` bash
NLOPT_LIBS = `${R_HOME}/bin/Rscript -e "nloptr:::LdFlags()"`
CXX_STD = CXX11
PKG_CXXFLAGS = $(SHLIB_OPENMP_CXXFLAGS)
PKG_LIBS = $(SHLIB_OPENMP_CFLAGS) $(LAPACK_LIBS) $(BLAS_LIBS) $(FLIBS) $(NLOPT_LIBS)
```
However my example package only installs when a system nlopt is present. Otherwise, I get a share library problem  (even if the Flags seem correct)

```
> devtools::load_all()
Loading nloptCpp
Re-compiling nloptCpp
'/usr/lib/R/bin/R' --no-site-file --no-environ --no-save --no-restore --quiet CMD INSTALL  \
  '/home/jchiquet/git/RcppArmadilloNLoptExample/nloptCpp' --library='/tmp/RtmpHFSSoy/devtools_install_31c5bc83b8e'  \
  --no-R --no-data --no-help --no-demo --no-inst --no-docs --no-exec --no-multiarch --no-test-load --preclean 

* installing *source* package ‘nloptCpp’ ...
** libs
g++ -std=gnu++11 -I/usr/share/R/include -DNDEBUG  -I"/home/jchiquet/R/x86_64-pc-linux-gnu-library/3.4/Rcpp/include" -I"/home/jchiquet/R/x86_64-pc-linux-gnu-library/3.4/RcppArmadillo/include" -I"/home/jchiquet/R/x86_64-pc-linux-gnu-library/3.4/nloptr/include"   -fopenmp -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g -c RcppExports.cpp -o RcppExports.o
g++ -std=gnu++11 -I/usr/share/R/include -DNDEBUG  -I"/home/jchiquet/R/x86_64-pc-linux-gnu-library/3.4/Rcpp/include" -I"/home/jchiquet/R/x86_64-pc-linux-gnu-library/3.4/RcppArmadillo/include" -I"/home/jchiquet/R/x86_64-pc-linux-gnu-library/3.4/nloptr/include"   -fopenmp -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g -c nlopt_cpp.cpp -o nlopt_cpp.o
g++ -std=gnu++11 -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -o nloptCpp.so RcppExports.o nlopt_cpp.o -fopenmp -llapack -lblas -lgfortran -lm -lquadmath -lm -L/usr/lib/R/lib -lR
installing to /tmp/RtmpHFSSoy/devtools_install_31c5bc83b8e/nloptCpp/libs
* DONE (nloptCpp)
Error in dyn.load(dllfile) : 
  unable to load shared object '/home/jchiquet/git/RcppArmadilloNLoptExample/nloptCpp/src/nloptCpp.so':
  /home/jchiquet/git/RcppArmadilloNLoptExample/nloptCpp/src/nloptCpp.so: undefined symbol: nlopt_set_munge
```

I don't know if it might be the reason but when installing nloptr without nlopt system library installed, I get the following messages at some point, even if the installation succeeds

``` bash
make[2]: *** No rule to make target 'nlopt-guile.cpp', needed by 'all'.  Stop.
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
make[1]: *** No rule to make target 'nlopt-guile.cpp', needed by 'install'.  Stop.
make: *** [install-recursive] Error 1
./configure: line 3348: checking: command not found
configure: Done installing library to /tmp/RtmphFnBVI/devtools17b07ea5bd83/jyypma-nloptr-eecb1aa/src/nlopt_src
mkdir: cannot create directory 'inst/include': File exists
```

The full log of nloptr install is [available here](https://github.com/jchiquet/RcppArmadilloNLoptExample/blob/master/nloptCpp/inst/nloptr_install.log)

**Comment by eddelbuettel** [comment]
> However my example package only installs when a system nlopt is present.

Yes. We went over this:
- your C example will be completely and reliably served by nlopt 1.1.0 or later
- your C++ example _will always need an external libnlopt_

They key is that _even if nloptr downloads nlopt and builds it, it simple 'slurps' it in as a static library_. It is not exported for general use. You cannot link against it as you could with a system library because _the host system does not see it_ only the nloptr build system does.

**Comment by jchiquet** [comment]
> They key is that even if nloptr downloads nlopt and builds it, it simple 'slurps' it in as a static library. It is not exported for general use. You cannot link against it as you could with a system library because the host system does not see it only the nloptr build system does.

Ok, I understand now. I was not aware of that point. Thanks, I stop bothering.

**Comment by svilsen** [comment]
> They key is that even if nloptr downloads nlopt and builds it, it simple 'slurps' it in as a static library. It is not exported for general use. You cannot link against it as you could with a system library because the host system does not see it only the nloptr build system does.

A follow-up question: Would it be possible for nloptr to export the downloaded and compiled nlopt library? If no, how come? If yes, would it be difficult?

**Comment by eddelbuettel** [comment]
@svilsen You cannot from an R package (an application) install and export a system library (such as upstream nltop).  That, sadly, remains OS-dependent and is out of the hands of R (or any other application as no other application language does it).  

All you can do is bundle it in your package (as nloptr does) and re-export.  Which we did here.  There really is no other "free lunch".

**Comment by svilsen** [comment]
@eddelbuettel Understood. I will 'borrow' the configure scripts from the nloptr package then. Thank you for your time.

**Comment by eddelbuettel** [comment]
Follow-up -- at long last, and thanks to a _lot_ of patience by @jyypma the new version 1.2.0 is now on CRAN.

Which means we can now deploy the _fabulous_ API provided by this version we can hit from C code, C++ code, RcppCode and hence also RcppArmadillo, RcppEigen, ... code -- and all this requires nothing more than `nloptr` being installed.   I crafted [a first example package RcppNLoptExample](https://github.com/eddelbuettel/rcppnloptexample) which extends the work I refered to above in a fork of the repo by @jchiquet, but simplifies it further.  Comments welcome!   

And again thanks to @jyypma for making this possible.
